### DISCUSSION TITLE
Assignment 1 is due on September 28 (Extended: October 4)

### DISCUSSION THREAD

**Discussion by ibnesayeed** [cs531-f22/cs531-f22.github.io#13] [Assignment]
[Assignment 1](https://cs531-f22.github.io/assignments/assignment-1/) is due on `2022-09-28`. This is the first main assignment in which you are asked to implement a basic HTTP server that supports a limited set of methods, status codes, and headers as described in the assignment details. Read related RFCs and refer to the first two lectures as needed.

The series of five main assignments will ask you to gradually add standard-compliant features in your HTTP server. Architect your system wisely and avoid hard-coding of configurable aspects to make it easier to extend your project for future assignments.

Feel free to post any questions or concerns you might have regarding this assignment in this thread.

**Comment by ibnesayeed** [comment]
The deadline of [Assignment 1](https://cs531-f22.github.io/assignments/assignment-1/) is extended to October 2.

**Comment by ibnesayeed** [comment]
If you are unsure how to include the test files in your server when deploying in Docker, you may try the following steps:

* Download the test tarball from https://raw.githubusercontent.com/ibnesayeed/webserver-tester/master/sample/cs531-test-files.tar.gz on your machine and move it in a folder named `sample` (or choose any other name as you prefer) inside of your code's root folder, without extracting the `tar.gz` file (because if you extract it, then empty folders will not make it to git repo and file permissions will also be messed up)

* Add the following line in your `Dockerfile` (change `sample` to something else, if you chose a different folder name in the previous step)

```
ADD     sample/cs531-test-files.tar.gz /var/www/
```

This will automatically extract the TAR file inside of `/var/www/` folder of the Docker image and preserve the directory structure

* Finally, use `/var/www/` as your WEBROOT in your server, so that if you receive a request for the `/a1-test/2/index.html` file, read it from `/var/www/a1-test/2/index.html` location instead

**Comment by calebkbrad** [comment]
I have a question about payload generation. When the target resource for a GET request has an image MIME type, what would be the expected payload from that request? Would it just be the raw bytes of that image?

**Comment by ibnesayeed** [comment]
> Would it just be the raw bytes of that image?

Yes, serve the raw bites of all static files, irrespective of the content type.

**Reply by calebkbrad** [reply]
Ok, thank you so much!

**Comment by rohitrahiya** [comment]
Hello Prof.,

I have created my server and exposed it at 3001 using docker run command with -p option of port mapping.

docker run --rm -it -p 80:3001 -v ${PWD}:/app --entrypoint bash echo:v1.0

In another terminal I ran the webserver tester docker container with below commands:
docker image build -t webserver-tester .
docker container run --rm -it --entrypoint bash webserver-tester 

Since, I am on Windows machine it does not allow me to directly run the main.py script.

Within the docker web-tester container, i run

python3 main.py 0.0.0.0:3001

But, I get connection refused error. I even tried local host.

My server runs at 0.0.0.0:80 with 3001 port mapping in docker run.

Can you help me with setting up tester application locally.

![image](https://user-images.githubusercontent.com/113197553/193392401-611e592e-9097-4c47-8ae7-201d6693db9e.png)

Regards,
Rohit

**Comment by ibnesayeed** [comment]
@rohitrahiya if you are running both your server and the tester in containers then they will not be able to talk to each other unless they share the same docker network (which is something you will have to create in advance and associate those containers with that network when running) and then you can access those containers by their private IPS, names, or IDs as hostnames.

As I explained during the socket programming lecture that the network interface `0.0.0.0` is not a resolvable address, it can only be used for listening to all interfaces, but not by the client (in this case, the tester) for connecting to the server.

The tester application has no dependencies other than the core Python packages, so you might just want to run it on the host machine (if you have Python installed), irrespective of whether your own server is running inside a container or on the host. Note that if your server is running inside of a container and you are using the port mapping `80:3001`, then you will be connecting to the port `80` on localhost from the host machine, not the port `3001` (the latter is internal to the container).

Another possibility would be to run a container from Python image with your code mounted inside of it and execute bash shell. Then run your server script from inside (not at the time of initializing the container). This way, you can use it like a regular Linux terminal that will allow to install packages and perform any other tasks you need to make your server work. Then, from another terminal, you can use `docker container exec <container_id> bash` command to create another terminal that is connected to your previous container, and run tests from there (you will have to download/copy the tester script inside).

Alternatively, you can create a Git branch like `assign1` and push your work in progress code on GitHub to test it on our testing server. Then make changes, push those again, and test again.

**Comment by ibnesayeed** [comment]
Considering recent power outages in some areas due to the bad weather the deadline of [Assignment 1](https://cs531-f22.github.io/assignments/assignment-1/) is extended to October 4.

**Comment by yaswanth9909** [comment]
Hello professor ,there's a problem with webserver testing .I am able to deploy the server but the test cases are not running .But when I run it in local host I can see the output .
![webserver](https://user-images.githubusercontent.com/73889665/193487657-645e57ee-4237-4eb0-8771-e0b01399facf.PNG)
![image](https://user-images.githubusercontent.com/73889665/193487667-896f5125-2870-4c67-bc9d-538e5276123e.png)

**Comment by ibnesayeed** [comment]
@yaswanth9909 the tester was failing to parse your response for the `cs531a1: test_get_jpeg_image` test case because you are returning binary data (i.e., the JPEG image) without the empty line after headers to separate it from the response headers. I have added some logic in the tester to detect such issues in the future (https://github.com/ibnesayeed/webserver-tester/pull/54/files).

Your implementation has many other issues, of which, some are detected by the tester.

**Comment by yaswanth9909** [comment]
Hello professor, I think there's a problem with the testing link
![image](https://user-images.githubusercontent.com/73889665/193961558-e08d8526-d4fe-4a49-960a-f8f35faab701.png)

**Comment by Badjedi04** [comment]
Hello Professor, 
The Tester seems to be down. I'm attaching the error screenshot.
I was in the process of testing my server and I realized that the Tester stopped working.
Could you please look into this.
![Tester Error](https://user-images.githubusercontent.com/98077449/193961780-91973861-570b-4704-b2ad-aae2c4eabd32.jpg)

**Comment by JMNewsome** [comment]
I am having trouble with the testing being down as well

**Comment by ibnesayeed** [comment]
The site should be up now. It looks like too many people testing their services at the same time caused it to reach the memory limit, resulting in application crashes. I will increase resources for the next assignment.

Sorry for the inconvenience. Submit your assignment by 8 AM, October 5, to avoid late penalties.
